cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+* Any two `null` values are treated as equivalent (both directly or inside nested structures).
+* Any two `NaN` values are treated as equivalent (both directly or inside nested structures).
+
+Note that orderability is aligned with equivalence, i.e. equivalent values have the same position under orderability.
+
+Equivalence is reflexive for all values.
+
+=== Summary of Conceptual Model
+
+This proposal aims to simplify the conceptual model around equality, comparison, sorting, and grouping:
+
+* Equality: Equality follows natural, literal equality. However, values involving `null` are never equal. Nested structures are first tested for equality by shape (keys, size) and then componentwise. This ensures that equality is compatible with interpreting `null` as "unknown" or "could be any value".
+* Comparability: Comparability is aligned with equality. Two values of the same type in the global sort order are always comparable. Two values of different types are always incomparable and values involving `null` are always incomparable.
+This ensures that `MATCH (n) WHERE n.prop > 42` will never find nodes where `n.prop` is of type `STRING`.
+* Equivalence: Equivalence is a form of equality that treats `null` (and `NaN`) values as identical. Equivalence is used in grouping and `DISTINCT` where `null` commonly is interpreted as a marker that a value was missing instead of as a wildcard for any possible value.
+* Orderability follows comparability but additionally defines a global sort order between values of different types and is aligned with equivalence regarding the handling of `null` and `NaN` values.
+
+=== Examples
+
+An integer compared to a float
+[source, cypher]
+----
+RETURN 1 > 0.5 // should be true
+----
+
+A string compared to a boolean
+[source, cypher]
+----
+RETURN 'string' <= true // should be null
+----
+
+Ordering values of different types
+[source, cypher]
+----
+UNWIND [1, true, '', 3.14, {}, [2]] AS i
+RETURN i
+  ORDER BY i // should not crash
+----
+
+Filtering distinct values of different types
+[source, cypher]
+----
+UNWIND [[null], [null]] AS i
+RETURN DISTINCT i // should return exactly one row
+----
+
+=== Interaction with existing features
+
+The concept of orderability is used only by `ORDER BY` in Cypher today.
+The concept of comparability is used by the comparisons operators `<`, `>`, `<=`, >=`.
+
+One major goal of our proposal is for equality semantics to align well with comparability.
+The concept of equality is used by the equality operator `=`, the inequality operator `<>`, value joins, and the `IN` operator.
+The concept of equivalence is used by the `DISTINCT` clause modifier and in grouping.
+
+With the proposals made in this CIP, specifically changing equality for lists, the mentioned functionality is going to treat lists containing `null` as unequal, thus potentially filtering out more rows when used in a predicate.
```

Comment:
This is a very, very, very long sentence. 

"Owing to this proposal altering equality semantics for lists - in particular, treating lists containing `null` as unequal - potentially more rows may be filtered out when used in a predicate." or something along these lines...